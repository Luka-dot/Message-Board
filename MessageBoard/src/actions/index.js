// actions
export const fetchNotes = (allNotes) => ({
  type: 'FETCH_NOTES',
  allNotes
})

export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const isLoading = (value) => ({
  type: 'IS_LOADING',
  value
}) 
