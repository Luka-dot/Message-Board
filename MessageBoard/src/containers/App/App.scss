.App {
  box-sizing: border-box;
  min-height: 100vh;
  padding: 0;
  margin: 0;
  text-align: center;

  .Content-Container {
    width: 100%;
    max-width: 1400px;
    margin: auto;
  }
}
