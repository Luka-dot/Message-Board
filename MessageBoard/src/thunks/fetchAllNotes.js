import { fetchNotes, hasError, isLoading } from '../actions';

export const fetchAllNotes = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url);
      dispatch(isLoading(false));
      if(!response.ok){
        throw Error(response.statusText)
      }
      const result = await response.json();
      dispatch(fetchNotes(result));
    } catch (error) {
      dispatch(hasError(error.message));
    }
  }
}
