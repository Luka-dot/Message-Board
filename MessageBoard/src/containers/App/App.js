import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom'
import { connect } from 'react-redux';
import { hasError } from '../../actions/index';
import { Header } from '../../components/Header/Header';
import NotesContainer from '../NotesContainer/NotesContainer';
import NoteForm from '../NoteForm/NoteForm';
import PropTypes from 'prop-types'
import { fetchAllNotes } from '../../thunks/fetchAllNotes'
import { ErrorPage } from '../../components/ErrorPage/ErrorPage'

export class App extends Component {

  componentDidMount() {
    const url = 'http://localhost:3001/api/v1/notes'
    this.props.fetchAllNotes(url)
  }

  render() {
    return (
      <div className="App">
        <Header />
        <main className="Content-Container">
        <Switch>
            <Route path="/" exact render={() => {
              return <NotesContainer />
            }} />
            <Route path="/notes/:id" render={({match}) => {
              const { id } = match.params;
              const currentNote = this.props.allNotes.find(note => note.id === id);
              if(currentNote === undefined) return <ErrorPage />
              return <NoteForm type="existing-note" noteId={id} {...currentNote}/>
            }} />
            <Route path="/new-note" render={ () => <NoteForm type="new-note"/> } />
            <Route component={ErrorPage} />
          </Switch>
        </main>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  allNotes: state.allNotes,
  error: state.error
})

export const mapDispatchToProps = (dispatch) => ({
  showError: (message) => dispatch(hasError(message)),
  fetchAllNotes: (url) => dispatch(fetchAllNotes(url)),
})

App.propTypes = {
  allNotes: PropTypes.array,
  error: PropTypes.string,
  showError: PropTypes.func,
  fetchNotes: PropTypes.func
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
