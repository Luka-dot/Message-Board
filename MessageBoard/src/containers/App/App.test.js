import React from 'react';
import { shallow, mount } from 'enzyme'
import { App, mapStateToProps, mapDispatchToProps } from './App';

import { hasError } from '../../actions'
import { fetchAllNotes } from '../../thunks/fetchAllNotes'


jest.mock('../../thunks/fetchAllNotes');

describe('App', () => {

  describe('App Component', () => {
    let wrapper;
    let mockfetchAllNotes;

    const mockAllNotes = [
      { id: "123ABA",
      title: 'Worf ToDo',
      list: [
        { id: "123213", text: 'Eat food', isComplete: false },
      ]
      },
      { id: "987GDGFD",
      title: 'Jake ToDo',
      list: [
        { id: "2353534543", text: 'Do basic styling', isComplete: false },
      ]
      }
    ];

    beforeEach(() => {
      mockfetchAllNotes = jest.fn().mockImplementation(() => Promise.resolve({results: mockAllNotes}))
      wrapper = shallow(<App fetchAllNotes={mockfetchAllNotes} />)
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it("should invoke the function fetchAllNotes on componentDidMount", async () => {
      wrapper.instance().componentDidMount()
      expect(mockfetchAllNotes).toBeCalled()
    })

    it("should call fetchAllNotes with expected url", async () => {
      const mockUrl = 'http://localhost:3001/api/v1/notes';
      wrapper.instance().componentDidMount()
      expect(mockfetchAllNotes).toHaveBeenCalledWith(mockUrl)
    })




  });

  describe('mapStateToProps', () => {
    it('should return an object with a notes array and an error string', () => {
      const mockState = {
        allNotes: [
          { id: "123ABA",
            title: 'Worf ToDo',
            list: [
              { id: "123", text: 'Eat food', isComplete: false },
            ]
          },
          { id: "987GDGFD",
            title: 'Jake ToDo',
            list: [
              { id: "0098", text: 'Do styling', isComplete: false },
            ]
          }
        ],
        error: "",
        fakeState: "Not real state to return"
      }
      const expected = {
        allNotes: [
          { id: "123ABA",
            title: 'Worf ToDo',
            list: [
              { id: "123", text: 'Eat food', isComplete: false },
            ]
          },
          { id: "987GDGFD",
            title: 'Jake ToDo',
            list: [
              { id: "0098", text: 'Do styling', isComplete: false },
            ]
          }
        ],
        error: "",
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using showError function from MDTP', () => {
      const mockDispatch = jest.fn()
      const mockError = "Something went wrong"
      const actionToDispatch = hasError(mockError)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.showError(mockError)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch on fetchAllNotes from MDTP', () => {
      const mockDispatch = jest.fn()
      const mockUrl = "www.googlenotes.com"
      const actionToDispatch = fetchAllNotes(mockUrl)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.fetchAllNotes(mockUrl)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  });

});
