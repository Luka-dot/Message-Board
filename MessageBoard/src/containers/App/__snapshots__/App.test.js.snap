// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App Component should match the snapshot 1`] = `
<div
  className="App"
>
  <Header />
  <main
    className="Content-Container"
  >
    <Switch>
      <Route
        exact={true}
        path="/"
        render={[Function]}
      />
      <Route
        path="/notes/:id"
        render={[Function]}
      />
      <Route
        path="/new-note"
        render={[Function]}
      />
      <Route
        component={[Function]}
      />
    </Switch>
  </main>
</div>
`;
